packages/library/src/commonMain/kotlin/io/realm/Realm.kt

```diff
@@ -57,10 +64,54 @@ class Realm private constructor(configuration: RealmConfiguration, dbPointer: Na
         this(configuration, RealmInterop.realm_open(configuration.nativeConfig))
 
     /**
-     * TODO Add docs when this method is implemented.
+     * Modify the underlying Realm file in a suspendable transaction on the default Realm
+     * dispatcher.
+     *
+     * NOTE: Objects and results retrieved before a write are no longer valid. This restriction
+     * will be lifted when the frozen architecture is fully in place.
+     *
+     * The write transaction always represent the latest version of data in the Realm file, even if
+     * the calling Realm not yet represent this.
+     *
+     * Write transactions automatically commit any changes made when the closure returns unless
+     * [MutableRealm.cancelWrite] was called.
+     *
+     * @param block function that should be run within the context of a write transaction.
+     * FIXME Isn't this impossible to achieve in a way where we can guarantee that we freeze all
+     *  objects leaving the transaction? It currently works for RealmObjects, and can maybe be done
+     *  for collections and RealmResults, but what it is bundled in other containers, etc. Should
+     *  we define an interface of _returnable_ objects that follows some convention?
+     * @return any value returned from the provided write block as frozen/immutable objects.
+     * @see [RealmConfiguration.writeDispatcher]
      */
-    suspend fun <R> write(block: MutableRealm.() -> R): R {
-        TODO("Awaiting implementation of the Frozen Architecture")
+    // TODO Would we be able to offer a per write error handler by adding a CoroutineExceptinoHandler
+    internal suspend fun <R> write(block: MutableRealm.() -> R): R {
+        @Suppress("TooGenericExceptionCaught") // FIXME https://github.com/realm/realm-kotlin/issues/70
+        try {
+            val (nativePointer, versionId, result) = this.writer.write(block)
+            // Update the user facing Realm before returning the result.
+            // That way, querying the Realm right after the `write` completes will return
+            // the written data. Otherwise, we would have to wait for the Notifier thread
+            // to detect it and update the user Realm.
+            updateRealmPointer(nativePointer, versionId)
+            // FIXME What if the result is of a different version than the realm (some other
```

Comment:
I guess it makes most sense to return the explicit version from the write block. But maybe we should eliminate the return value for now as there are quite more to it if we want to ensure that anything returned here is actually frozen. Alternatively we should try to limit `R` to some set of `Freezable` entities to make it explicit that it only works for `RealmObject` and `RealmResult` that we can immediately freeze

Reply:
I left the freezing at the version of the write block, but have still restricted it to only freeze the result if it is a `RealmObject`. Don't know if we should throw if it is not the case?

Reply:
The user facing `Realm` is advancing as it is always pointing to the latest state we have, but any objects obtained from it will be frozen to the version.